### DISCUSSION TITLE
What is the point in auto pairs for a function's brackets `()`?

### DISCUSSION THREAD

**Discussion by David-Else** [helix-editor/helix#4051] [Q&A] [answered]
If I am creating a function I get `()` with the cursor inside. I always have to press a sequence of characters or simply the closing bracket to leave and continue... so why bother with auto pairs when using the closing bracket was the best option anyway? If we could tab out it would be fine, but that is not the case.

I can see it is great for things like `{}` as you don't need to skip over the closing pair all the time.

**Comment by dead10ck** [comment, marked as answer]
Many functions have arguments.

For functions that don't, most keyboard layouts have `)` right next to `(`, so I think for most people just typing `()` is a quick roll of two fingers. In fact, I'd think that doing `(<tab>` would be way more awkward.

**Reply by David-Else** [reply]
I have not tried Helix with auto-pairs off yet, could you be more explicit in what you mean?

**Reply by dead10ck** [reply]
For functions that have parameters, it is most convenient to close the parens and leave the cursor inside so you can just start typing the function arguments.

For functions that don't, you can just type `()`; helix will just skip over the `)` instead of inserting another one.

**Reply by David-Else** [reply]
I am not sure why I need the brackets closed to enter the arguments, would I miss out on LSP functionality?

**Reply by dead10ck** [reply]
Well there are potentially technical reasons, like tree-sitter wouldn't recognize it as a function call until the closing paren was there, so it may end up colored incorrectly, or possibly it might cause issues with some of the tree-sitter navigation movements.

But I'm not really sure I understand your line of thought, is there some problem? If there were no auto pairs at all, you would type `function_name(` and either start typing arguments, or a closing `)`. And if you have auto pairs enabled, you would... type `function_name(` and either start typing arguments, or a closing `)`, and get the same result.

Other times, it could potentially save you a key stroke if the function call was the last thing you needed to do on that line. It's just a small convenience. One might ask the same question about auto pairs in general.

**Reply by kirawi** [reply]
@pickfire doesn't use auto-pairs, so it's fine if you don't like the behavior.

**Reply by dead10ck** [reply]
Also, if you'd really prefer not to have `()` auto paired, but would still like everything else, this is possible. You could follow the instructions [here](https://docs.helix-editor.com/master/configuration.html#editorauto-pairs-section) and leave `()` out.

**Reply by David-Else** [reply]
Nicely explained @dead10ck
> Other times, it could potentially save you a key stroke if the function call was the last thing you needed to do on that line

I was not really thinking about that situation

**Reply by pickfire** [reply]
> @pickfire doesn't use auto-pairs, so it's fine if you don't like the behavior.

Yup, I don't use auto-pairs, auto-completion, auto-format (I like auto-format only for rust). I rather manually type the pair as I go rather than quit insert mode and move over the pair. Also, the pair doesn't really work well in the case you type your argument over spaces IIRC

```rust
fn t(
    a: A,
) <-- this
```

But I noticed doom emacs does solve this issue such that if you type `)` even though it's not your next character, it will jump over the auto-inserted `)`.
